bcipy/language/model/gpt2.py

```diff
@@ -0,0 +1,188 @@
+        """
+        assert self.response_type == ResponseType.SYMBOL, "response type must be SYMBOL!"
+
+    def __get_char_predictions(self, word_prefix: str) -> List[tuple]:
+        """
+        Given a word prefix typed by user, predict the probability distribution
+        of next character
+        Args:
+            word_prefix - string
+        Response:
+            A list of tuples - each tuple consists a pair of symbol and probability
+        """
+
+        prefix_len = len(word_prefix)
+        char_to_prob = Counter()
+
+        # iterate over the entire vocabulary
+        for idx in range(self.vocab_size):
+            # change word and word_prefix to upper case
+            word = self.idx_to_word[idx].upper()
+            word_prefix = word_prefix.upper()
+
+            # if a word in the vocabulary matches the word prefix
+            if word.startswith(word_prefix):
+                # obtain the character following the prefix in this word
+                if len(word) == prefix_len:
+                    next_char = SPACE_CHAR
+                else:
+                    next_char = word[prefix_len]
+                # make sure the next character is in the symbol set
+                if next_char in self.symbol_set and next_char != BACKSPACE_CHAR:
+                    char_to_prob[next_char] += self.curr_word_predicted_prob[idx]
+
+        # normalize char_to_prob
+        sum_char_prob = sum(char_to_prob.values())
+        for char in char_to_prob:
+            char_to_prob[char] /= sum_char_prob
+
+        # build a list of tuples (char, prob)
+        char_prob_tuples = [(k, v) for k, v in sorted(char_to_prob.items(), key=lambda item: item[1], reverse=True)]
+
+        return char_prob_tuples
+
+    def __build_vocab(self) -> Dict[int, str]:
+        """
+        Build a vocabulary table mapping token index to word strings
+        Response:
+            a dict that maps token index to word strings
+        """
+
+        idx_to_word = {}
+        for idx in range(self.vocab_size):
+            word = self.tokenizer.decode([idx]).strip()
+            idx_to_word[idx] = word
+
+        return idx_to_word
+
+    def __model_infer(self, text: str) -> List[float]:
+        """
+        Use the transformer language model to infer the distribution of next word
+        Args:
+            text - a string of text representing the text input so far
+        Response:
+            An numpy array representing next word's probability distribution over
+            the entire vocabulary
+        """
+
+        # empty text indicates the beginning of a phrase, attach the BOS symbol
+        if text == "":
+            text = "<|endoftext|>"
+
+        # tokenize the text
+        indexed_tokens = self.tokenizer.encode(text)
+        tokens_tensor = torch.tensor([indexed_tokens])
+
+        # predict the next word
+        with torch.no_grad():
+            model = self.model
+            outputs = model(tokens_tensor)
```

Comment:
I don't think the `model` variable here is needed.

    self.model(tokens_tensor)

Reply:
Yes, I will change this.